### ISSUE TITLE
ArgumentError on range iteration

### ISSUE THREAD

**Issue by pawelpacana** [oracle/truffleruby#1503]
I was evaluating my code on truffle ruby and noticed following behavior:

```ruby 
 Code $ ruby -e 'p (0..Float::INFINITY).step(2).take(10)'
-e:1:in `floor': Infinite or NaN (ArgumentError)
	from -e:1:in `step_float_iterations_size'
	from -e:1:in `step_internal'
	from -e:1:in `step'
	from -e:1:in `__send__'
	from -e:1:in `each_with_block'
	from -e:1:in `each'
	from -e:1:in `take'
	from -e:1:in `<main>'

 Code $ ruby -v                                                                                                                              
truffleruby 1.0.0-rc9, like ruby 2.4.4, GraalVM CE Native [x86_64-darwin]
```

For the reference this MRI:

```
 Code $ ruby -e 'p (0..Float::INFINITY).step(2).take(10)'
[0.0, 2.0, 4.0, 6.0, 8.0, 10.0, 12.0, 14.0, 16.0, 18.0]

 Code $ ruby -v
ruby 2.5.3p105 (2018-10-18 revision 65156) [x86_64-darwin18]
```

And this is JRuby:

```
 Code $ ruby -e 'p (0..Float::INFINITY).step(2).take(10)'
[0.0, 2.0, 4.0, 6.0, 8.0, 10.0, 12.0, 14.0, 16.0, 18.0]

 Code $ ruby -v                                                                                                                        
jruby 9.2.4.0 (2.5.0) 2018-11-13 8faff06 Java HotSpot(TM) 64-Bit Server VM 25.152-b16 on 1.8.0_152-b16 +jit [darwin-x86_64]
```

Is that expected?

I've finally changed my code to avoid unwanted floats which resulted in:

```
 Code $ ruby -v -e 'p 0.step(Float::INFINITY, 2).each.take(10)'
ruby 2.5.3p105 (2018-10-18 revision 65156) [x86_64-darwin18]
[0, 2, 4, 6, 8, 10, 12, 14, 16, 18]


 Code $ ruby -v -e 'p 0.step(Float::INFINITY, 2).each.take(10)'
jruby 9.2.4.0 (2.5.0) 2018-11-13 8faff06 Java HotSpot(TM) 64-Bit Server VM 25.152-b16 on 1.8.0_152-b16 +jit [darwin-x86_64]
[0, 2, 4, 6, 8, 10, 12, 14, 16, 18]
```

This however does not look good on truffle:

```
 Code $ ruby -v -e 'p 0.step(Float::INFINITY, 2).each.take(10)'
truffleruby 1.0.0-rc9, like ruby 2.4.4, GraalVM CE Native [x86_64-darwin]
[]
```

Related: https://github.com/RailsEventStore/rails_event_store/commit/a5d5153815f70b4527e2d034ec0926a470dcb92c + https://github.com/RailsEventStore/rails_event_store/commit/626c5708aa20a765f362c4207847e62f29a4d046

**Comment by chrisseaton** [comment]
Looks like we have inherited both these bugs from Rubinius. Hopefully an easy fix.

**Comment by deepj** [comment]
@eregon @chrisseaton Can you add this as a priority to RC13. I've seen this issue several times in some tests/specs.

**Comment by eregon** [comment]
@nirvdrum Could you look at this since you worked on `step` recently?

**Comment by nirvdrum** [comment]
Will do. At first glance, it's been broken so at least I didn't regress. But we probably need a new Ruby spec.

**Comment by nirvdrum** [comment]
I believe I've fixed this issue in b53957f31c8d4336806bfd9deab00869c9b97673. The fix should appear in the 1.0.0 RC 13 release.
